test/functional/tests/fault/test_primary_device_error.py

```diff
@@ -0,0 +1,155 @@
+        description: |
+          Perform I/O on two exported objects created using error and non-error device.
+          Validate CAS that stats counting is consistent with OS reporting.
+          Also, check if normal I/O is uninterrupted and no DC occurs on any of the
+          core devices.
+        pass_criteria:
+          - I/O error count in FIO and in cache statistics match
+          - Positively passed fio verify on both core devices
+    """
+    cache, error_core, good_core = prepare_configuration(cache_mode, cache_line_size)
+
+    good_core_fio = (
+        Fio()
+        .create_command()
+        .io_engine(IoEngine.libaio)
+        .size(good_core.size)
+        .block_size(cache_line_size)
+        .target(good_core)
+        .read_write(ReadWrite.randrw)
+        .verify_pattern()
+        .verify(VerifyMethod.pattern)
+        .direct()
+    )
+
+    error_core_fio = (
+        Fio()
+        .create_command()
+        .io_engine(IoEngine.libaio)
+        .size(error_core.size)
+        .block_size(cache_line_size)
+        .target(error_core)
+        .read_write(io_dir)
+        .continue_on_error(ErrorFilter.io)
+        .direct()
+    )
+    if io_dir == ReadWrite.randwrite:
+        error_core_fio = error_core_fio.verify_pattern().verify(VerifyMethod.pattern)
+
+    with TestRun.step("Run fio on core without errors in background"):
+        fio_pid = good_core_fio.run_in_background()
+
+    fio_errors = 0
+    with TestRun.step("Run fio on error core and check if IO errors are present"):
+        fio_errors = error_core_fio.run()[0].total_errors()
+
+        if fio_errors == 0:
+            TestRun.fail("No I/O ended with error!")
+
+    with TestRun.step("Check error statistics on error core"):
+        stats = cache.get_statistics()
+
+        core_errors_in_cache = stats.error_stats.core.total
+        if fio_errors != core_errors_in_cache:
+            TestRun.fail(
+                f"Core errors in cache stats({core_errors_in_cache}) should equal to number of"
+                " fio errors ({fio_errors})"
+            )
+
+    with TestRun.step("Wait for fio on good core"):
+        TestRun.executor.wait_cmd_finish(fio_pid)
+
+    with TestRun.step("Check error statistics on good core"):
+        stats = good_core.get_statistics()
+
+        if stats.error_stats.core.total != 0:
+            TestRun.fail(
+                f"No errors should be reported for good core. "
+                "Actual result: {stats.error_stats.total}"
+            )
+
+    with TestRun.step("Stop the cache"):
+        cache.stop()
+
+    if io_dir == ReadWrite.randwrite:
+        with TestRun.step("Verify error core device contents"):
+            error_core_fio.target(error_core.core_device).verify_only().run()
+
+    with TestRun.step("Verify good core device contents"):
+        good_core_fio.target(good_core.core_device).verify_only().run()
+
+
+def prepare_configuration(cache_mode, cache_line_size):
```

Comment:
I thought that for the purpose of parsing test steps, we weren't going to enclose steps within methods

Reply:
Fixed, also removed conditional step to make it easier